### PR TITLE
prov/efa: cleanup configure.m4

### PR DESCRIPTION
**Pull request by lrbison**
Address some conflicting styles and inconsistent macro usage in
configure.m4.

This commit doesn't have impactful changes for efa configuration.

### CODE REVIEW

**Comment by lrbison** [comment]
@bwbarrett, what do you think about this from a style point of view?

**Comment by wenduwan** [comment]
This change is actually touching a lot of features... We probably need to look into the CI and make sure we didn't disable(skip) any test.

File: prov/efa/configure.m4 [hunk 1]
```diff
@@ -151,40 +148,32 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
+	AC_DEFINE_UNQUOTED([HAVE_EFA_DATA_IN_ORDER_ALIGNED_128_BYTES],
+		[$efa_support_data_in_order_aligned_128_byte],
+		[Indicates if EFA supports 128 bytes in-order in writing.])
 
-	AS_IF([test $efa_support_data_in_order_aligned_128_byte -eq 1],
-		[AC_DEFINE([HAVE_EFA_DATA_IN_ORDER_ALIGNED_128_BYTES], [1], [EFA device supports 128 bytes in-order in writing.])],
-		[AC_DEFINE([HAVE_EFA_DATA_IN_ORDER_ALIGNED_128_BYTES], [0], [EFA device does not support 128 bytes in-order in writing.])]
-		)
 
 	efa_CPPFLAGS="$efa_ibverbs_CPPFLAGS $efadv_CPPFLAGS"
```
Comment (bwbarrett) [inline prov/efa/configure.m4:178]: I would put these further down, after all the tests (even though they're unit test related tests, someone else might miss that distinction).

**Review by bwbarrett — CHANGES_REQUESTED** [review]

File: prov/efa/configure.m4 [hunk 2]
```diff
@@ -45,92 +45,91 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
+		AC_CHECK_DECL([IBV_QUERY_QP_DATA_IN_ORDER_ALIGNED_128_BYTES],
+			[efa_support_data_in_order_aligned_128_byte=1],
+			[efa_support_data_in_order_aligned_128_byte=0],
+			[[#include <infiniband/verbs.h>]])
 
-	AS_IF([test $ac_cv_sizeof_void_p -eq 4],
-		[
-			efa_happy=0
-			AC_MSG_WARN([The EFA provider is not supported on 32-bit systems.])
+		dnl Check for ibv_reg_dmabuf_mr() in libibverbs if built with synapseai support.
```
[the remark this answers is not in the document]
Comment (shijin-aws) [reply prov/efa/configure.m4:118]: Yes, neuron team actually already merged a PR to support dmabuf in libfabric: https://github.com/ofiwg/libfabric/pull/8517

File: prov/efa/configure.m4 [hunk 3]
```diff
@@ -45,92 +45,91 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
+			[-libverbs],
+			[$efa_PREFIX],
+			[$efa_LIBDIR],
+			[],
+			[
+				efa_happy=0
+				AC_MSG_WARN([The EFA provider requires rdma-core v31 or newer.])
+			])
+
+		FI_CHECK_PACKAGE([efadv],
```
[the remark this answers is not in the document]
Comment (lrbison) [reply prov/efa/configure.m4:65]: I'm not sure... we did this already (https://github.com/ofiwg/libfabric/blob/main/prov/efa/configure.m4#L63) and the difference is the first time we check for `efadv_query_ah` and print a warning, then we check for `efadv_query_device` and don't print a warning if it isn't found....  I didn't check to see if that made sense, just duplicated it.

File: prov/efa/configure.m4 [hunk 4]
```diff
@@ -45,92 +45,91 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
-			      [[#include <infiniband/efadv.h>]])
-	      ])
 
-	AS_IF([test x"$enable_efa" != x"no"],
-	      [AC_CHECK_DECL(EFADV_DEVICE_ATTR_CAPS_RNR_RETRY,
-			    [AC_DEFINE([HAVE_CAPS_RNR_RETRY], [1], [EFADV_DEVICE_ATTR_CAPS_RNR_RETRY is defined])],
-			    [AC_DEFINE([HAVE_CAPS_RNR_RETRY], [0], [EFADV_DEVICE_ATTR_CAPS_RNR_RETRY is not defined])],
-			    [[#include <infiniband/efadv.h>]])
-	      ])
+	dnl define these defaults for when efa_happy is 0
```
[the remark this answers is not in the document]
Comment (lrbison) [reply prov/efa/configure.m4:84]: yes, agreed

File: prov/efa/configure.m4 (hunk 2 above)
[the remark this answers is not in the document]
Comment (lrbison) [reply prov/efa/configure.m4:118]: I see.  Sounds like what you are proposing is to introduce a new HAVE_IBV_REG_DMA_BUF_MR.  I think that should go in a separate PR.

One thing I realized in all this is that these names we are using here (such as HAVE_IBV_IS_FORK_INITIALIZED) are in fact global to all of libfabric.  That makes me slightly uncomfortable that there might be an unintentional name collision.  I'm interested in renaming all of these `HAVE_EFA_....` or something like that as well.

**Comment by lrbison** [comment]
> This change is actually touching a lot of features... We probably need to look into the CI and make sure we didn't disable(skip) any test.

Don't I know it!
I did a few variations with enable/disable some things, and compared my config.h output.  I can't claim full coverage but at least I gave it a sniff test before pushing.

File: prov/efa/configure.m4 (hunk 3 above)
[the remark this answers is not in the document]
Comment (lrbison) [reply prov/efa/configure.m4:65]: I checked, and query_ah is slightly newer (2019 May vs April), so I kept only that check.

File: prov/efa/configure.m4 (hunk 1 above)
Comment (lrbison) [reply prov/efa/configure.m4:156]: I'm not sure which lines you are referring to here.

File: prov/efa/configure.m4 [hunk 5]
```diff
@@ -45,92 +45,91 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
 		])
 
-	AS_IF([test $efa_happy -eq 1 ], [$1], [$2])
-
-	AS_IF([test $efadv_support_extended_cq -eq 1],
-		[FI_CHECK_PACKAGE([efadv],
+		dnl For efadv_support_extended_cq, we check several things,
+		dnl and if any of them fail, we disable CQ_EX
+		efadv_support_extended_cq=1
+		FI_CHECK_PACKAGE([efadv],
```
[the remark this answers is not in the document]
Comment (lrbison) [reply prov/efa/configure.m4]: Done.  Changed these all to AC_CHECK_DECL

**Review by bwbarrett — APPROVED** [review]

File: prov/efa/configure.m4 [hunk 6]
```diff
@@ -45,195 +41,160 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
 			[],
-			[efa_support_data_in_order_aligned_128_byte=0])
-		])
+			[efadv_support_extended_cq=0],
+			[[#include <infiniband/efadv.h>]])
+		AC_CHECK_DECL([efadv_cq_from_ibv_cq_ex],
+			[],
+			[efadv_support_extended_cq=0],
+			[[#include <infiniband/efadv.h>]])
+		AC_CHECK_DECL([efadv_wc_read_sgid],
```
Comment (shijin-aws) [inline prov/efa/configure.m4:129]: I would let @wenduwan have a comment here. I think he uses EGREP_HEADER for some reason. Not sure AC_CHECK_DECL works here

**Review by wenduwan — APPROVED** [review]
I was mostly concerned with features being disabled by this change. Looking at the test report it looks ok though - the skipped unit tests should be expected.
https://libfabric-ci.aws.a2z.com/job/multi-node/381/PROVIDER=efa,label=alinux2/testReport/alinux2.test_suites.libfabric/test_efa_ut/

File: prov/efa/configure.m4 (hunk 6 above)
Comment (wenduwan) [reply prov/efa/configure.m4:129]: I checked the unit test report and the tests were run so I think it worked.

**Pull request merged**